### PR TITLE
GRW-393 / Feature / Tabs

### PR DESCRIPTION
**Pull request by dendoesit**
GRW-393 / Feature / Tabs

## What?

Added a new tabs components to storybook with an underline that moves based on the selected item. 
The tabs can be scrolled on both mobile and desktop. 
Updated framer-motion to 4.x and all components that use it.
Updated Embark to 2.2.0

## Why?

We want to stay up-to-date with the latest design changes in Figma.

Referenced ticket(s): [GRW-393]

image

[GRW-393]: https://hedvig.atlassian.net/browse/GRW-393?
Desktop
atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Mobile with scroll

image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -134,7 +134,7 @@
     "escape-html": "^1.0.3",
     "eventemitter3": "^4.0.7",
     "formik": "^2.0.8",
-    "framer-motion": "^1.6.15",
+    "framer-motion": "^4.1.17",
```
Comment (robinandeer) [inline package.json:137]: Have you checked that everything is working correctly after this update?

I had to tweak a few animation configs after doing a similar update in Embark I remember...

File: src/client/pages/OfferNew/Tabs/Underline.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import { motion } from 'framer-motion'
+import { MEDIUM_SCREEN_MEDIA_QUERY } from 'utils/mediaQueries'
+
+const Underline = styled(motion.div)`
+  width: 100%;
+  position: absolute;
+  top: 1.9rem;
+  ${MEDIUM_SCREEN_MEDIA_QUERY} {
+    top: 2.125rem;
+  }
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Underline.tsx:12]: Do we really need these non-standard spacing values?

File: src/client/pages/OfferNew/Tabs/Tabs.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import React, { useState } from 'react'
+import { colorsV3 } from '@hedviginsurance/brand'
+import { AnimateSharedLayout } from 'framer-motion'
+import styled from '@emotion/styled'
+import { MEDIUM_SCREEN_MEDIA_QUERY } from 'utils/mediaQueries'
+import { Tab } from './Tab'
+
+const TabGroup = styled.div`
+  border-bottom: 1px solid ${colorsV3.gray500};
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tabs.tsx:9]:
```suggestion
  border-bottom: 2px solid ${colorsV3.gray500};
```

...however, from Figma it seems like this is implemented as a box-shadow, have you tried how that looks?

```
box-shadow: 0px 2px 0px #EAEAEA;
```

File: src/client/pages/OfferNew/Tabs/Tabs.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  return (
+    <TabGroup>
+      <AnimateSharedLayout>
+        {insurances.map((item, index) => (
+          <Tab
+            key={index}
+            onClick={() => setSelected(item)}
+            selected={selected === item}
+            name={item.name}
+          ></Tab>
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tabs.tsx:38]:
```suggestion
          <Tab
            key={index}
            onClick={() => setSelected(item)}
            selected={selected === item}
            name={item.name}
          />
```

File: src/client/pages/OfferNew/Tabs/Tab.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  focused?: boolean
+}>`
+  color: ${({ selected }) => (selected ? colorsV3.gray900 : colorsV3.gray500)};
+  cursor: pointer;
+  display: flex;
+  font-size: 1.25rem;
+  line-height: 1.75rem;
+  padding-bottom: 0.25rem;
+  position: relative;
+  text-transform: capitalize;
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tab.tsx:18]: Is this necessary?

File: src/client/pages/OfferNew/Tabs/Tab.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        inline: 'center',
+      })
+    }
+  }
+
+  return (
+    <TabContainer
+      onClick={() => scroll()}
+      selected={selected}
+      ref={ref as React.MutableRefObject<HTMLDivElement>}
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tab.tsx:56]: Why the need to downcast here?

File: src/client/pages/OfferNew/Tabs/Tab.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+  const scroll = () => {
+    onClick()
+    if (ref.current) {
+      ref.current.scrollIntoView({
+        behavior: 'smooth',
+        block: 'nearest',
+        inline: 'center',
+      })
+    }
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tab.tsx:49]: What is this supposed to do? Do we want to scroll the page when you tab so another section?

File: src/client/pages/OfferNew/Tabs/Tabs.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+export const Default = () => {
+  const [insurances, setInsurances] = React.useState(mockInsurances)
+
+  const handleOnChange = (id: string) =>
+    setInsurances((prev) =>
+      prev.map((item) => ({ ...item, selected: item.id === id })),
+    )
+
+  return <Tabs insurances={insurances} onChange={handleOnChange} />
+}
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tabs.stories.tsx:38]: Perhaps show here also how we can use this setup to show different content based on selected tab?

File: src/client/pages/OfferNew/Tabs/Tab.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import React from 'react'
+import styled from '@emotion/styled'
+import { colorsV3 } from '@hedviginsurance/brand'
+import { MEDIUM_SCREEN_MEDIA_QUERY } from 'utils/mediaQueries'
+import { UnderlineComponent } from './Underline'
+
+const TabContainer = styled.div<{
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tab.tsx:7]: If this is "clickable" it should be a `button` element :)

**Review by robinandeer — CHANGES_REQUESTED** [review]
Nice work! Sorry for all the comments ;)

Overall, I would like us to invest some time to use standard markup as far as possible. One of my go-to sources is Reach UI: https://reach.tech/tabs/

Maybe we don't want to pull in a dependency but will should still aim to follow standards I believe :)

File: src/client/pages/OfferNew/Tabs/Tabs.tsx [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    margin-top: 1rem;
+    overflow: visible;
+  }
+`
+
+type Props = {
+  insurances: {
+    id: string
+    name: string
+    content?: string
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tabs.tsx:28]: Is this representing what to display when the tab is selected?

I think we need something more flexible like `React.ReactNode`, right?
Also, should it be optional?

File: src/client/pages/OfferNew/Tabs/Tabs.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        <AnimateSharedLayout>
+          {insurances.map((item, index) => (
+            <>
+              <Tab
+                key={index}
+                onClick={() => setSelected(item)}
+                selected={selected === item}
+                name={item.name}
+              />
+            </>
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tabs.tsx:46]:
```suggestion
            <Tab
              key={index}
              onClick={() => setSelected(item)}
              selected={selected === item}
              name={item.name}
            />
```

File: src/client/pages/OfferNew/Tabs/Tabs.tsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+import { Tab } from './Tab'
+
+const TabList = styled.div`
+  display: flex;
+  margin-bottom: 1.25rem;
+  margin-right: 1rem;
+  margin-top: 1rem;
+  max-width: 100%;
+  overflow-x: scroll;
+  -webkit-overflow-scrolling: touch;
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Tabs.tsx:14]: I think this has been removed from webkit. Do we really need it?

https://caniuse.com/mdn-css_properties_-webkit-overflow-scrolling

File: src/client/pages/OfferNew/Tabs/Underline.tsx [hunk 13]
```diff
@@ -0,0 +1,28 @@
+import React from 'react'
+import styled from '@emotion/styled'
+import { motion } from 'framer-motion'
+import { colorsV3 } from '@hedviginsurance/brand'
+import { MEDIUM_SCREEN_MEDIA_QUERY } from 'utils/mediaQueries'
+
+const Underline = styled(motion.div)`
+  box-shadow: 0px 2px 0px ${colorsV3.gray900};
```
Comment (robinandeer) [inline src/client/pages/OfferNew/Tabs/Underline.tsx:8]: Did you get this to work with box-shadow? :D

**Review by robinandeer — COMMENTED** [review]
Supernice update!

It's seems so generic I'm intrigued to rename the "insurances" prop to like "items" to go all the way ;)

File: src/client/pages/OfferNew/Tabs/Underline.tsx (hunk 13 above)
Comment (dendoesit) [reply src/client/pages/OfferNew/Tabs/Underline.tsx:8]: this one yes. but the Tab needs border-bottom :(

**Comment by robinandeer** [comment]
Something is strange about the underline animation when switching from tab 2 to tab 1... no idea what's going on though 🤔 

https://user-images.githubusercontent.com/1220232/137888207-547f6790-3494-4c6d-9907-feb076ed2b38.mov

**Review by robinandeer — COMMENTED** [review]
Cool stuff! I really have only one thing to note about using our predefined media queries :)

Otherwise looks good to me - great work!

**Review by robinandeer — COMMENTED** [review]
Have you looked into @gustaveen comment about underline floating on top of the shadow/border?

Otherwise I think I'm happy with the API/PR now :)

**Review by robinandeer — DISMISSED** [review]
I think we are ready to merge this now -- great work! :D

`Tabs/index.tsx` seems empty tho so should we remove it perhaps?

**Review by robinandeer — APPROVED** [review]
re-approve

**Pull request merged**